### PR TITLE
Overhaul sequence filename definitions 

### PR DESCRIPTION
**Pull request by pchote**
This is the first in a series of PRs aiming to remove technical debt from the sequence code and add features needed to finish and upstream remastered asset support.

All the magic around sprite filenames is removed in favour of explicitly specifying the full sprite name (per tileset if needed). This allows mods that don't follow the existing C&C conventions (e.g. by putting each tilesets assets in its own directory) to use tileset-specific artwork. It also prepares for a future PR that will add new ways of specifying filenames (e.g. `FilenamePattern: mytank-{:D4}.png`).

The `Defaults` node was previously combined into each sequence using `MiniYaml.Merge` at sequence-parsing time. This prevented us from implementing recursive merging at the `MiniYaml` level (see e.g. [this complaint](https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Game/GameRules/WeaponInfo.cs#L141)) because the raw sequence yaml may be malformed (e.g. trying to remove keys that don't exist) before the defaults have been copied over. This has been replaced with a much simpler fallback convention: the parser first looks for the requested key in the current sequence, and if not found it will look again in `Defaults`. Sequences can override (or more correctly, mask) the `Defaults` value by defining a key with no value.

I suggest reviewing by commit, and reading the commit descriptions for more information.
The yaml changes in the third commit are fully automated, so it shouldn't be necessary to inspect every single line in detail. A small number of manual cleanups are specified in the fourth commit.

### CODE REVIEW

File: OpenRA.Mods.Common/UpdateRules/UpdateUtils.cs [hunk 1]
```diff
@@ -196,9 +228,19 @@ public static List<string> UpdateMod(ModData modData, UpdateRule rule, out YamlF
 			}
 
 			manualSteps.AddRange(rule.BeforeUpdate(modData));
+
+			var resolvedActors = MiniYaml.Load(modData.DefaultFileSystem, modData.Manifest.Rules, null);
+			manualSteps.AddRange(rule.BeforeUpdateActors(modData, resolvedActors));
 			manualSteps.AddRange(ApplyTopLevelTransform(modData, modRules, rule.UpdateActorNode));
+
+			var resolvedWeapons = MiniYaml.Load(modData.DefaultFileSystem, modData.Manifest.Weapons, null);
+			manualSteps.AddRange(rule.BeforeUpdateSequences(modData, resolvedWeapons));
```
Comment (PunkPun) [inline OpenRA.Mods.Common/UpdateRules/UpdateUtils.cs]:
```suggestion
			manualSteps.AddRange(rule.BeforeUpdateWeapons(modData, resolvedWeapons));
```

File: OpenRA.Mods.Common/UpdateRules/UpdateUtils.cs [hunk 2]
```diff
@@ -114,6 +114,9 @@ public static List<string> UpdateMap(ModData modData, IReadWritePackage mapPacka
 				var mapRulesNode = yaml.Nodes.FirstOrDefault(n => n.Key == "Rules");
 				if (mapRulesNode != null)
 				{
+					var resolvedActors = LoadMapYaml(modData.DefaultFileSystem, mapPackage, modData.Manifest.Rules, mapRulesNode.Value);
+					manualSteps.AddRange(rule.BeforeUpdateWeapons(modData, resolvedActors));
```
Comment (PunkPun) [inline OpenRA.Mods.Common/UpdateRules/UpdateUtils.cs]:
```suggestion
					manualSteps.AddRange(rule.BeforeUpdateActors(modData, resolvedActors));
```

**Review by PunkPun — COMMENTED** [review]
Since we are changing everything up we may want a more user friendly exception for missing values. It doesn't mention where something is missing nor what

I deleted the filename field from MiG and got this
```
Exception of type `System.IO.FileNotFoundException`: File not found: 
   at OpenRA.Mods.Common.Graphics.FileNotFoundSequence.OpenRA.Graphics.ISpriteSequence.get_Tick() in /Users/gustas/Documents/GitHub/OpenRA/OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs:line 112
   at OpenRA.Graphics.Animation.PlaySequence(String sequenceName) in /Users/gustas/Documents/GitHub/OpenRA/OpenRA.Game/Graphics/Animation.cs:line 126
   ```
   
   Completely unhelpful to modders

**Comment by PunkPun** [comment]
The CheckSequences lint test (or a new one) could also be used to pick up on missing values

**Comment by pchote** [comment]
Fixed.

**Comment by PunkPun** [comment]
This is much better 👍. However it would be more ideal if lint test didn't display the tracelogs

**Comment by pchote** [comment]
The next PR completely overhauls the way that the sprite loading and lint tests work, so can we leave that for now?

**Comment by PunkPun** [comment]
sure

**Comment by PunkPun** [comment]
I tried updating [Treasure_Hunt.oramap.zip](https://github.com/OpenRA/OpenRA/files/10300023/Treasure_Hunt_BI-3.5-2.oramap.zip) and the update rule crashed

<details>
  <summary>Crashlog</summary>

```
System.Collections.Generic.KeyNotFoundException: The given key 'SNOW' was not present in the dictionary.
        at OpenRA.Mods.Common.UpdateRules.Rules.ExplicitSequenceFilenames.ProcessNode(ModData modData, MiniYamlNode sequenceNode, MiniYamlNode resolvedSequenceNode, String imageName) in /Users/gustas/Documents/GitHub/OpenRA/OpenRA.Mods.Common/UpdateRules/Rules/20221203/ExplicitSequenceFilenames.cs:line 281
        at OpenRA.Mods.Common.UpdateRules.Rules.ExplicitSequenceFilenames.UpdateSequenceNode(ModData modData, MiniYamlNode imageNode)+MoveNext() in /Users/gustas/Documents/GitHub/OpenRA/OpenRA.Mods.Common/UpdateRules/Rules/20221203/ExplicitSequenceFilenames.cs:line 136
        at OpenRA.Mods.Common.UpdateRules.UpdateUtils.ApplyTopLevelTransform(ModData modData, List`1 files, TopLevelNodeTransform transform)+MoveNext() in /Users/gustas/Documents/GitHub/OpenRA/OpenRA.Mods.Common/UpdateRules/UpdateUtils.cs:line 292
        at System.Collections.Generic.List`1.InsertRange(Int32 index, IEnumerable`1 collection)
        at System.Collections.Generic.List`1.AddRange(IEnumerable`1 collection)
        at OpenRA.Mods.Common.UpdateRules.UpdateUtils.UpdateMap(ModData modData, IReadWritePackage mapPackage, UpdateRule rule, List`1& files, HashSet`1 externalFilenames) in /Users/gustas/Documents/GitHub/OpenRA/OpenRA.Mods.Common/UpdateRules/UpdateUtils.cs:line 143
        at OpenRA.Mods.Common.UtilityCommands.UpdateMapCommand.ApplyRules(ModData modData, IReadWritePackage mapPackage, IEnumerable`1 rules) in /Users/gustas/Documents/GitHub/OpenRA/OpenRA.Mods.Common/UtilityCommands/UpdateMapCommand.cs:line 108
```
</details>

**Comment by pchote** [comment]
The update rule requires the legacy tileset metadata (which is removed in the last commit) to be defined in mod.yaml. The map updates fine if you add that back.

**Comment by PunkPun** [comment]
Seems less than ideal. Update rules are also meant for maps, I'd even say that they're used for maps more than for mods. We should handle the crashes

**Comment by pchote** [comment]
I can’t think of an acceptable way to do that - keeping unused junk data in mod.yaml isn’t acceptable, and hardcoding the upgrade rule with data for only the upstream mods isn’t really either.

File: OpenRA.Mods.Common/UpdateRules/Rules/20221203/ExplicitSequenceFilenames.cs [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+					if (!tilesetOverrides.TryGetValue(tileset, out var sequenceTileset))
+						sequenceTileset = tileset;
+
+					var overrideFilename = filename;
+					if (useTilesetCode)
+						overrideFilename = filename.Substring(0, 1) + tilesetCodes[sequenceTileset] +
+						                   filename.Substring(2, filename.Length - 2);
+
+					if (addExtension)
+						overrideFilename += useTilesetExtension ? tilesetExtensions[sequenceTileset] : defaultSpriteExtension;
```
Comment (PunkPun) [inline OpenRA.Mods.Common/UpdateRules/Rules/20221203/ExplicitSequenceFilenames.cs:387]: We could just check if the `sequenceTileset` exists and if it doesn't just not do the step. Instead we could log a note for the modder to fix it manually

Comment (PunkPun) [reply OpenRA.Mods.Common/UpdateRules/Rules/20221203/ExplicitSequenceFilenames.cs:387]: The same issue could happen with TilesetCodes as well correct?

**Comment by abcdefg30** [comment]
The update rule shouldn't crash but write a proper error message - explaining what is written in the comments above: That it needs the definitions in mod.yaml and that those can be removed (again) afterwards.

**Comment by pchote** [comment]
Updated.

```
ExplicitSequenceFilenames: Sequence filenames must be specified explicitly.
   Updating map... COMPLETE
   Manual changes are required to complete this update:
    * The ExplicitSequenceFilenames rule requires TilesetExtensions
      to be defined under the SpriteSequenceFormat definition in mod.yaml.
      Add these definitions back and run the update rule again.

Semi-automated update complete.
Please review the messages above for any manual actions that must be applied.
```

**Comment by PunkPun** [comment]
In the future we should make it clearer when update rules fail / require manual changes. Currently that information is fairly hard to find within walls of text.

I did run this rule on some other fairly artwork heavy maps and it seems to function well

**Review by PunkPun — DISMISSED** [review]
LGTM

**Comment by PunkPun** [comment]
Could there be anything done to the resource centre update function to account for the missing definitions?

**Comment by pchote** [comment]
Yes, the definitions can be added back to mod.yaml on the RC engines.

File: OpenRA.Mods.Cnc/Graphics/ClassicSpriteSequence.cs [hunk 4]
```diff
@@ -20,9 +20,9 @@ public class ClassicSpriteSequenceLoader : DefaultSpriteSequenceLoader
 		public ClassicSpriteSequenceLoader(ModData modData)
 			: base(modData) { }
 
-		public override ISpriteSequence CreateSequence(ModData modData, string tileSet, SpriteCache cache, string sequence, string animation, MiniYaml info)
+		public override ISpriteSequence CreateSequence(ModData modData, string tileset, SpriteCache cache, string image, string sequence, MiniYaml data, MiniYaml defaults)
```
Comment (penev92) [inline OpenRA.Mods.Cnc/Graphics/ClassicSpriteSequence.cs:23]: I am not convinced `sequence` -> `image` and `animation` -> `sequence` makes it better. My recommendation for "SequenceCollection" + "Sequence" (or "Animation"?) still stands.
To me *"image"* was the problem we had before, not the solution.

**Comment by PunkPun** [comment]
Needs a rebase / copyright header update

**Comment by pchote** [comment]
Rebased.

File: mods/ts/sequences/vehicles.yaml [hunk 5]
```diff
@@ -1,189 +1,242 @@
...
-	icon: cyclicon
+	icon:
+		Filename: cyclicon.shp
 
 bggy:
 	Inherits: ^VehicleOverlays
 	Defaults:
 		Offset: 0, 0, 24
 	muzzle:
+		Filename:
```
Comment (Mailaender) [inline mods/ts/sequences/vehicles.yaml]: Why is this required?

File: OpenRA.Mods.Cnc/Graphics/ClassicSpriteSequence.cs [hunk 6]
```diff
@@ -33,15 +33,13 @@ public class ClassicSpriteSequence : DefaultSpriteSequence
...
+		public ClassicSpriteSequence(ModData modData, string tileSet, SpriteCache cache, ISpriteSequenceLoader loader, string image, string sequence, MiniYaml data, MiniYaml defaults)
+			: base(modData, tileSet, cache, loader, image, sequence, data, defaults)
 		{
-			var d = info.ToDictionary();
-			useClassicFacings = LoadField(d, UseClassicFacings);
+			useClassicFacings = LoadField(UseClassicFacings, data, defaults);
 
 			if (useClassicFacings && facings != 32)
-				throw new InvalidOperationException(
-					$"{info.Nodes[0].Location}: Sequence {sequence}.{animation}: UseClassicFacings is only valid for 32 facings");
```
Comment (Mailaender) [inline OpenRA.Mods.Cnc/Graphics/ClassicSpriteSequence.cs:44]: Why are you removing the line number?

File: OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs [hunk 7]
```diff
@@ -41,49 +41,49 @@ public bool TryGetPaletteForFrame(int frame, out uint[] palette)
...
 			{
-				throw new InvalidDataException($"Error occurred while parsing {node.Key}", e);
+				throw new InvalidDataException($"Error occurred while parsing {imageNode.Key}", e);
 			}
 
-			foreach (var kvp in nodes)
+			foreach (var sequenceNode in imageNode.Value.Nodes)
 			{
-				using (new Support.PerfTimer($"new Sequence(\"{node.Key}\")", 20))
+				using (new Support.PerfTimer($"new Sequence(\"{imageNode.Key}\")", 20))
```
Comment (penev92) [inline OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs:71]: Unrelated, but isn't this supposed to be *outside* of the `foreach`?
Also unrelated, but the `return new ReadOnlyDictionary<string, ISpriteSequence>(sequences);` at the end is redundant. Should just be `return sequences;` and the method's signature will ensure consumers only see an `IReadOnlyDictionary<>`.

File: OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs [hunk 8]
```diff
@@ -414,7 +417,10 @@ public DefaultSpriteSequence(ModData modData, string tileSet, SpriteCache cache,
 						return subFrames != null ? subFrames.Skip(subStart).Take(subLength) : Enumerable.Range(subStart, subLength);
 					};
 
-					var subFilename = GetSpriteFilename(modData, tileSet, combineSequenceNode.Key, sequence, combineData, NoData);
+					var subFilename = GetSpriteFilename(modData, tileSet, image, sequence, combineData, NoData);
+					if (subFilename == null)
+						throw new YamlException($"Sequence {image}.{sequence}.{combineSequenceNode.Key} does not define a filename");
```
Comment (penev92) [inline OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs]: Full stop at the end of the sentence please.

File: OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs [hunk 9]
```diff
@@ -440,6 +446,8 @@ public DefaultSpriteSequence(ModData modData, string tileSet, SpriteCache cache,
 				// Apply offset to each sprite in the sequence
 				// Different sequences may apply different offsets to the same frame
 				var filename = GetSpriteFilename(modData, tileSet, image, sequence, data, defaults);
+				if (filename == null)
+					throw new YamlException($"Sequence {image}.{sequence} does not define a filename");
```
Comment (penev92) [inline OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs]: Same + empty line below.

File: OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs [hunk 10]
```diff
@@ -497,6 +505,9 @@ public DefaultSpriteSequence(ModData modData, string tileSet, SpriteCache cache,
 			if (LoadField(HasEmbeddedPalette, data, defaults))
 			{
 				var filename = GetSpriteFilename(modData, tileSet, image, sequence, data, defaults);
+				if (filename == null)
+					throw new YamlException($"Sequence {image}.{sequence} does not define a filename");
```
Comment (penev92) [inline OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs]: Full stop at the end of the sentence please.

File: OpenRA.Mods.Common/UpdateRules/Rules/20221203/ExplicitSequenceFilenames.cs [hunk 11]
```diff
@@ -0,0 +1,412 @@
...
+using System.Reflection;
+
+namespace OpenRA.Mods.Common.UpdateRules.Rules
+{
+	public class ExplicitSequenceFilenames : UpdateRule
+	{
+		public override string Name => "Sequence filenames must be specified explicitly.";
+
+		public override string Description =>
+			"Sequence sprite filenames are no longer automatically inferred, and the AddExtension.\n" +
```
Comment (penev92) [inline OpenRA.Mods.Common/UpdateRules/Rules/20221203/ExplicitSequenceFilenames.cs]:
```suggestion
			"Sequence sprite filenames are no longer automatically inferred and the AddExtension,\n" +
```

File: OpenRA.Mods.Common/UpdateRules/Rules/20221203/ExplicitSequenceFilenames.cs [hunk 12]
```diff
@@ -0,0 +1,412 @@
...
+			"UseTilesetExtension, UseTilesetNodes, and TilesetOverrides fields have been removed.\n\n" +
```
Comment (penev92) [inline OpenRA.Mods.Common/UpdateRules/Rules/20221203/ExplicitSequenceFilenames.cs]:
```suggestion
			"UseTilesetExtension, UseTilesetNodes and TilesetOverrides fields have been removed.\n\n" +
```

File: OpenRA.Mods.Common/UpdateRules/Rules/20221203/ExplicitSequenceFilenames.cs [hunk 13]
```diff
@@ -0,0 +1,412 @@
...
+			// Add Filename/TilesetFilenames nodes to every sequence
+			foreach (var sequenceNode in imageNode.Value.Nodes)
+			{
+				if (string.IsNullOrEmpty(sequenceNode.Key) || sequenceNode.KeyMatches("Inherits"))
+					continue;
+
+				if (!string.IsNullOrEmpty(sequenceNode.Value.Value))
+					explicitlyNamedSequences.Add(sequenceNode.Key);
+
+				var resolvedSequenceNode = resolvedImageNode.Value.Nodes.SingleOrDefault(n => n.Key == sequenceNode.Key);
```
Comment (penev92) [inline OpenRA.Mods.Common/UpdateRules/Rules/20221203/ExplicitSequenceFilenames.cs:169]: This will return `null` for node removals, causing an NRE in `ProcessNode` below. Please add a check above to ignore if this is a removal, like you do for "Inherits".

**Review by penev92 — CHANGES_REQUESTED** [review]
The naming may or may not just be me, but there is at least one crash to fix.

File: OpenRA.Mods.Cnc/Graphics/ClassicSpriteSequence.cs (hunk 4 above)
Comment (pchote) [reply OpenRA.Mods.Cnc/Graphics/ClassicSpriteSequence.cs:23]: `Sequence` and `Image` are the names used by traits. This PR is already big and rebase-prone, I *really* don't want to have to edit another 50+ trait files!

File: OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs (hunk 7 above)
Comment (pchote) [reply OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs:71]: These will be dealt with in #20597.

File: OpenRA.Mods.Cnc/Graphics/ClassicSpriteSequence.cs (hunk 6 above)
Comment (pchote) [reply OpenRA.Mods.Cnc/Graphics/ClassicSpriteSequence.cs:44]: `info.Nodes[0].Location` was often the wrong line! (and the current plumbing doesn't let us easily report the correct one)

This PR removes it for correctness and consistency with the other sequence yaml errors.
#20597 implements line numbers properly for all errors.

**Comment by PunkPun** [comment]
https://github.com/OpenRA/OpenRA/wiki/Sequences
will need to be updated after the merge

**Comment by pchote** [comment]
Updated.

**Review by Mailaender — DISMISSED** [review]
`--check-sprite-sequences` clean. No problems in-game.

File: OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs [hunk 14]
```diff
@@ -41,49 +41,49 @@ public bool TryGetPaletteForFrame(int frame, out uint[] palette)
 
 	public class DefaultSpriteSequenceLoader : ISpriteSequenceLoader
 	{
+		static readonly MiniYaml NoData = new MiniYaml(null);
```
Comment (penev92) [inline OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs:44]: This breaks `ExtractSpriteSequenceDocsCommand` ☹️ because apparently if you call `GetGenericTypeDefinition()` on a non-generic type it throws.
Please change
```diff
Properties = type.GetFields(BindingFlags.NonPublic | BindingFlags.Static)
-	.Where(fi => fi.FieldType.GetGenericTypeDefinition() == typeof(SpriteSequenceField<>))
+	.Where(fi => fi.FieldType.IsGenericType && fi.FieldType.GetGenericTypeDefinition() == typeof(SpriteSequenceField<>))
	.Select(fi =>
```
in `ExtractSpriteSequenceDocsCommand`, line 65.

**Comment by pchote** [comment]
Updated.

**Review by penev92 — APPROVED** [review]
👍

**Comment by penev92** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/1cd13a6e3b0d3f258483ba154af6962856106cc4...0ebf015d53056e47d3c82b836ce58e3266a81c50)
